storage/aws/aws.go

```diff
@@ -425,6 +426,180 @@ func (a *Appender) updateEntryBundles(ctx context.Context, fromSeq uint64, entri
+	logStore     *logResourceStore
+}
+
+var _ tessera.MigrationWriter = &MigrationStorage{}
+
+func (m *MigrationStorage) AwaitIntegration(ctx context.Context, sourceSize uint64) ([]byte, error) {
+	t := time.NewTicker(time.Second)
+	defer t.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-t.C:
+			from, _, err := m.sequencer.currentTree(ctx)
+			if err != nil && !errors.Is(err, os.ErrNotExist) {
+				klog.Warningf("readTreeState: %v", err)
+				continue
+			}
+			klog.Infof("Integrate from %d (Target %d)", from, sourceSize)
+			newSize, newRoot, err := m.buildTree(ctx, sourceSize)
+			if err != nil {
+				klog.Warningf("integrate: %v", err)
+			}
+			if newSize == sourceSize {
+				klog.Infof("Integrated to %d with roothash %x", newSize, newRoot)
+				return newRoot, nil
+			}
+		}
+	}
+}
+
+func (m *MigrationStorage) SetEntryBundle(ctx context.Context, index uint64, partial uint8, bundle []byte) error {
+	return m.logStore.setEntryBundle(ctx, index, partial, bundle)
+}
+
+func (m *MigrationStorage) IntegratedSize(ctx context.Context) (uint64, error) {
+	sz, _, err := m.sequencer.currentTree(ctx)
+	return sz, err
+}
+
+func (m *MigrationStorage) fetchLeafHashes(ctx context.Context, from, to, sourceSize uint64) ([][]byte, error) {
+	// TODO(al): Make this configurable.
+	const maxBundles = 300
+
+	toBeAdded := sync.Map{}
+	eg := errgroup.Group{}
+	n := 0
+	for ri := range layout.Range(from, to, sourceSize) {
+		eg.Go(func() error {
+			b, err := m.logStore.getEntryBundle(ctx, ri.Index, ri.Partial)
+			if err != nil {
+				return fmt.Errorf("getEntryBundle(%d.%d): %v", ri.Index, ri.Partial, err)
+			}
+
+			bh, err := m.bundleHasher(b)
+			if err != nil {
+				return fmt.Errorf("bundleHasherFunc for bundle index %d: %v", ri.Index, err)
+			}
+			toBeAdded.Store(ri.Index, bh[ri.First:ri.First+ri.N])
+			return nil
+		})
+		n++
+		if n >= maxBundles {
+			break
+		}
+	}
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+
+	lh := make([][]byte, 0, maxBundles)
+	for i := from / layout.EntryBundleWidth; ; i++ {
+		v, ok := toBeAdded.LoadAndDelete(i)
+		if !ok {
+			break
+		}
+		bh := v.([][]byte)
+		lh = append(lh, bh...)
+	}
+
+	return lh, nil
+}
+
+func (m *MigrationStorage) buildTree(ctx context.Context, sourceSize uint64) (uint64, []byte, error) {
+	var newSize uint64
+	var newRoot []byte
+
+	tx, err := m.dbPool.BeginTx(ctx, nil)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed to begin Tx: %v", err)
+	}
+	defer func() {
+		if tx != nil {
+			if err := tx.Rollback(); err != nil && err != sql.ErrTxDone {
+				klog.Errorf("failed to rollback Tx: %v", err)
+			}
+		}
+	}()
+
+	// Figure out which is the starting index of sequenced entries to start consuming from.
+	row := tx.QueryRowContext(ctx, "SELECT seq, rootHash FROM IntCoord WHERE id = ? FOR UPDATE", 0)
+	var from uint64
+	var rootHash []byte
+	if err := row.Scan(&from, &rootHash); err == sql.ErrNoRows {
+		return 0, nil, nil
```

Comment:
I think `initDB` (via `newMySQLSequencer`) should ensure there's a row in here, so this would represent a broken invariant - worth returning an error and/or logging I think.

Reply:
Made it fail whatever the error is.